Refactor pull query execution

vvcephei opened the pull request:
### Description 

* Consolidate ad-hoc query analysis 
* Clarify the type of query in various "execute query" methods in the codebase
* Misc refactors to prepare for stream pull queries

### Testing done 

* Existing tests should be sufficient for a refactor

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ImmutableAnalysis.java:
```diff
@@ -74,4 +75,6 @@
 
   boolean getOrReplace();
+
+  DataSource getDataSource();
```
```suggestion
  DataSource getFromDataSource();
```
in the spirit of bike-shedding, I feel like we shouldn't introduce this extra method since it makes the API a little confusing (when using autocomplete, I'll have to try to understand why we have `getFrom` and `getDataSource` and read the code to work out the difference - spoiler, there isn't one). But if we insist in saving the extra `.getDataSource()` call in the code base, let's try to make it clear that they're really the same method.

vvcephei replied:
Ah, I didn't know about `getFrom()`. I also just noticed that this method is unused in this PR, so I should have saved it for the next one. Sorry about that.

I just tried it out on the follow-on branch, and it looks like (Law of Demeter violation notwithstanding) I can just use `.getFrom().getDataSource()` instead.

vvcephei replied:
Hmm, now that I'm removing it, I notice that I'd written exactly that code in the implementation, so now I don't know _what_ I was thinking...

agavra replied:
😆

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java:
```diff
@@ -165,7 +162,8 @@ ExecuteResult execute(final KsqlPlan plan) {
    */
-  PullQueryResult executePullQuery(
+  PullQueryResult executeTablePullQuery(
+      final ImmutableAnalysis analysis,
```
leaving notes to make sure I understand the refactor. this is the "meat", right? we no longer do the analysis at each part of the query but rather ahead-of-time, which allows us to "understand" whether it's a table pull/stream pull before choosing the corresponding method?

vvcephei replied:
Yes!

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/QueryEndpoint.java:
```diff
@@ -133,19 +132,36 @@ public QueryPublisher createQueryPublisher(
+          requestProperties
+      );
     } else {
       return createPushQueryPublisher(context, serviceContext, statement, workerExecutor);
```
do non-scalable push queries not benefit from having the analysis also done up front? I was thinking why not do the analysis outside of the if/else tree?

vvcephei replied:
IIRC, the analysis currently needs to be a little different for transient and persistent queries, so I just didn't mess with it for now.

vvcephei replied:
After digging back through the code, I can confirm that was the reason, and I made another attempt to unify the paths, but changing transient queries has too much splash damage.

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryPublisher.java:
```diff
@@ -109,7 +110,11 @@ public synchronized void subscribe(final Subscriber<Collection<StreamedRow>> sub
     PullQueryResult result = null;
     try {
-      result = ksqlEngine.executePullQuery(
+      final ImmutableAnalysis analysis = ksqlEngine
```
why not pass the analysis into the constructor instead of doing it again here? it looks like you are creating it before creating the publisher in `QueryEndpoint`. Does it get lost before getting here? (this might be a different code path that I'm missing)

vvcephei replied:
I don't think there's anywhere before this point that we do compute the analysis. That said, we can certainly push it up more and further consolidate calls to the analyzer. I'll give it a shot.

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PushQueryPublisher.java:
```diff
@@ -126,12 +127,14 @@ public synchronized void subscribe(final Flow.Subscriber<Collection<StreamedRow>
               query.getSessionConfig().getConfig(false),
               query.getSessionConfig().getOverrides());
 
+      final ImmutableAnalysis analysis = ksqlEngine
+          .analyzeQueryWithNoOutput(query.getStatement(), query.getStatementText());
```
ditto from above, why not pass the analysis in?

vvcephei replied:
Oh, so I implemented this advice, but now we actually are computing the analysis twice for _transient_ queries. The reason is that this "push query publisher" has the same constructors whether you're doing a transient or scalable push queries. I tried to make it so we could pass the analysis into the transient query execution, but it turns into a real nightmare without completely disentangling it from persistent queries.

vvcephei replied:
Ok, I went ahead and partially walked this back, so for pull queries, we still do it at the top level for pull queries, but for push queries, we do the analysis as soon as we can distinguish between scalable vs. transient push queries.

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java:
```diff
@@ -460,18 +488,17 @@ private EndpointResponse handleScalablePushQuery(
     );
 
-    log.info("Streaming query '{}'", statement.getStatementText());
+    QueryLogger.info("Streaming query", statement.getStatementText());
```
putting some thoughts down here: we probably want to have another PR that goes through and standardizes what we do for query logging across the transient query types (and also logs the messages meaningfully, e.g. there's no indication that this is a scalable push query from the query log). doesn't belong in this PR though, can you create a ticket to track this (tag `fix-it-week` and maybe also `p1`)

vvcephei replied:
Sure, I agree we should audit the codebase and make sure we always use the logger for queries. I thought we already had, so I was surprised to find this here.

vvcephei replied:
Filed https://github.com/confluentinc/ksql/issues/7979

agavra reviewed (approved):
LGTM after rebasing with a green master

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/KsqlExecutionContext.java:
```diff
@@ -138,7 +139,7 @@
    * @return the query metadata
    */
-  TransientQueryMetadata executeQuery(
+  TransientQueryMetadata executeTransientQuery(
```
/cc @AlanConfluent but I don't think we're planning on entirely deprecating the traditional push queries in favor of scalable (e.g. you'll never be able to do a join/aggregate with a scalable push query IMO)

vvcephei replied:
Thanks, all. Regardless of our actual plans, just a little feedback on the feedback: the reason I wound up making these changes in this PR was that I had a really hard time tracking down what all the different "query" methods were actually doing. I think "execute old push query" is better than just "execute query", but it still requires you to know the context: that we plan to deprecate this query path, for example. Even then, there are any number of code paths we may plan to deprecate in the future, so "old" doesn't tell you which old query path this is for.

agavra replied:
+1 ^ thanks @vvcephei

vvcephei commented on ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java:
```diff
@@ -369,6 +383,20 @@ public static boolean isExecutableStatement(final Statement statement) {
         || statement instanceof Query;
   }
 
+  public ImmutableAnalysis analyzeQueryWithNoOutput(final Query query, final String queryText) {
```
I meant that the query is of a type that has no output topic. I guess I could burn an extra five characters and clarify that in the method name ;)

vvcephei commented on ksqldb-engine/src/main/java/io/confluent/ksql/engine/OrphanedTransientQueryCleaner.java:
```diff
@@ -37,7 +36,6 @@
   private final QueryCleanupService cleanupService;
   private final KsqlConfig ksqlConfig;
 
-  @SuppressFBWarnings(value = "EI_EXPOSE_REP")
```
Oh, this was probably some kind of rebase mistake.

vvcephei commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java:
```diff
@@ -334,65 +308,117 @@ private EndpointResponse handleStatement(
+      final Context context,
+      final SlidingWindowRateLimiter pullBandRateLimiter) {
+
+    if (statement.getStatement().isPullQuery()) {
```
That would be nice, but differentiating those actually requires the analysis. I did consider creating an AnalyzedStatement similar to ConfiguredStatement to just keep everything together, but it was too big of a change for right now.

vvcephei commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java:
```diff
@@ -334,65 +308,117 @@ private EndpointResponse handleStatement(
+      );
+    } else {
+      // log validated statements for query anonymization
+      QueryLogger.info("Transient query created", statement.getStatementText());
```
Hmm, I'll take a look.

vvcephei replied:
Ok, I've fixed the other log line to say "scalable"

vvcephei commented:
Thanks for the reviews, @agavra and @guozhangwang ! I've addressed all your comments and made another pass over the diff myself. I think this is ready to merge if/when the tests pass.
